### PR TITLE
Frontend passes target to backend

### LINKED ISSUE #78: Communication of playing a card with targets
For #41.

This requires there be an E2E test in place, and that the feature works as expected. This is the final step in implementation of issue #41.

### PR DESCRIPTION
**Pull request by Masclins**
This resolves #78.

But after this and #115 everything is **really** ugly.

Philosophy goes as following:
Cards in hand have an attribute `add-power` equal to its *add-power* ability or `"undefined"`.
When a card is "played", if `add-power === "undefined"` the play is passed to the backend, with `target: "nil"`. Otherwise it gets an attribute `row-played` with the row where it's been played.
When a card on the board is clicked with a card "played" (that means it must be a card with an `add-power !== "undefined"` previously dragged/clicked) the "played" card recieves an attribute `target: "[:rows 0 :cards 0]"` with 0's being the appropriate indexes (even I can see the ugliness in that).

On the other hand, backend recieves always a target, and `api.base/play-card-as-player` understand the target (a string) as a clojure code giving the path (using `(let [target (load-string target)]`.

From this pull-request (and the previous one) I see a couple of things that need work:
- We should decide how to handle targets. I coded everything regarding targets and most probably are suboptimal (I'm being really generous here).
- We are leaving a lot of *work* to the frontend, like deciding which cards need a target.
- Attribute-setting and calls to `play()` are done in many places on the frontend code. That's weird and messy.

### CODE REVIEW

File: backend/src/api/base.clj [hunk 1]
```diff
@@ -13,13 +13,16 @@
   (player-view/get-game-as-player (persistence/fetch-game game-id) player-id))
 
 (defn play-card-as-player
-  [game-id player index row-id & target]
-  (let [game-state (persistence/fetch-game game-id)]
+ ([game-id player index row-id]
+  (play-card-as-player game-id player index row-id "nil"))
+ ([game-id player index row-id target]
+  (let [game-state (persistence/fetch-game game-id)
+        target (load-string target)]
```
Comment (kenan-rhoton) [inline backend/src/api/base.clj]: `load-string` is a security issue, it is executing arbitrary Clojure code contained in a string received from an external connection.
A malicious user could alter the value being sent as `target` and run any kind of code on the system, compromising the entire backend.

File: frontend/tests/game/play.test.js [hunk 2]
```diff
@@ -12,11 +12,12 @@ test('Playing cards changes background as intended', () => {
     play.playCard(card);
     expect(card.style.background).toBe("green");
     
+    card = document.createElement("div");
     card.setAttribute("add-power", 1);
     play.playCard(card);
     expect(card.style.background).toBe("yellow");
 
-    card.setAttribute("target", 1);
+    card.setAttribute("target", "[:rows 0 :cards 0]");
```
Comment (kenan-rhoton) [inline frontend/tests/game/play.test.js]: We cannot send Clojure code. Either split the information into `target-row` and `target-card` or use a json-encoded object

**Review by kenan-rhoton — CHANGES_REQUESTED** [review]

**Comment by Masclins** [comment]
Now it passes the backend and frontend tests, but not the e2e.
So I think it's pretty obvious that the problem comes with the REST api.
Any help?

File: backend/src/api/base.clj (hunk 1 above)
Comment (kenan-rhoton) [reply backend/src/api/base.clj]: Just as a reminder: this pull request *cannot* be accepted until it doesn't use vulnerable things such as `load-string`

File: frontend/tests/game/play.test.js (hunk 2 above)
Comment (kenan-rhoton) [reply frontend/tests/game/play.test.js]: The removal of `load-string` makes this change now incorrect, right?

Comment (Masclins) [reply frontend/tests/game/play.test.js]: now it should pass the card-id

Comment (Masclins) [reply frontend/tests/game/play.test.js]: This test only checks the card has a target, it doesn't checks if the target is correct.

**Comment by kenan-rhoton** [comment]
@Masclins Just to confirm, you dropped this on purpose, right?

**Comment by Masclins** [comment]
I dropped this because branch was deleted and pushed unto your repository.
We can either merge the branch and @pernilsalat makes his changes on a new branch, or we wait until he improves this one.

**Pull request closed without merging**